### DISCUSSION TITLE
Span doesn't align with start and end id when '-' is performed in the text.

### DISCUSSION THREAD

**Discussion by effyli** [explosion/spaCy#7737] [Help: Coding & Implementations] [answered]
Currently I am using SpaCy visualizer and the Span class to visualize self-defined entity mentions. 
However I found out that when a token has '-' in it, the token will be split into three parts and therefore the visualization doesn't align with the ids anymore. 

## How to reproduce the behaviour
```
import spacy
from spacy import displacy
from spacy.tokens import Span

nlp = spacy.load("en_core_web_sm")
text = "A light earthquake has struck Northern California , and there are no reports of injuries or damage . The U.S. Geological Survey says the 4.4-magnitude temblor on Sunday at 9:27 a.m. was centered 23 miles north of Santa Rosa , and two miles from The Geysers . "
t = nlp(text)
print(t)
new_span = Span(t, 24, 25, "ent")
t.set_ents([new_span])
displacy.serve(t, style="ent")
```

As you can see in the result, only '4.4' is highlighted instead of '4.4-magnitude', which doesn't make sense to me.

**Comment by polm** [comment]
How did you generate your token IDs? It looks like you have ids from some other tool that don't align with spaCy tokens. When you create a span from 24 to 25 that just covers a single token, so everything here is working as expected.

What you can do is convert your token indexes to character indexes using your existing tokenizer, and then use [Span.char_span](https://spacy.io/api/span#char_span) to create spans. That would look like this:

```
new_span = Span.char_span(t, start_char, end_char, "ent")
t.set_ents([new_span])
displacy.serve(t, style="ent")
```

**Comment by adrianeboyd** [comment, marked as answer]
Alternatively, if you want to modify the `en_core_web_sm` tokenizer to tokenize on spaces since your text looks like it's pretokenized, you can modify the tokenizer settings as follows:

```python
import spacy
from spacy.tokens import Span
nlp = spacy.load("en_core_web_sm")
nlp.tokenizer.prefix_search = None
nlp.tokenizer.suffix_search = None
nlp.tokenizer.infix_finditer = None
nlp.tokenizer.token_match = None
nlp.tokenizer.url_match = None
nlp.tokenizer.rules = {}
text = "A light earthquake has struck Northern California , and there are no reports of injuries or damage . The U.S. Geological Survey says the 4.4-magnitude temblor on Sunday at 9:27 a.m. was centered 23 miles north of Santa Rosa , and two miles from The Geysers . "
t = nlp(text)
new_span = Span(t, 24, 25, "ent")
assert new_span.text == "4.4-magnitude"
```

If the tokenizer doesn't find any affixes, matches, or special cases (rules), then it splits the text on space characters. Be aware that whitespace other than a single space will lead to additional whitespace tokens, which would throw off your token offsets, so if you use this method, be sure your input text only uses a single space to separate tokens.

**Reply by effyli** [reply]
Thanks for the reply. This seems very helpful :).
